### DISCUSSION TITLE
Wet bulb temperature

### DISCUSSION THREAD

**Discussion by bonnells** [open-meteo/open-meteo#1091] [Ideas]
I am aware of a previous topic referring to showers and whether they are rain or snow. The most appropriate parameter to delineate between the two is the wet-bulb temperature. This is not the only area where wet-bulb temperatures are useful in providing value in the forecasting arena; another focus is on heat stress. Given the increasing importance of this aspect of modern day forecasts, I strongly recommend that those in this community become more aware of the relevance of the wet-bulb temperature to any output they wish to derive for their users.

It is my intention to make a few comments on the rain/snow issue but in advance of doing this, I'd like to suggest that either Patrick, at his initial involvement with the data, or any of you subsequently deriving forecasts from this, implement some code to derive this variable.

The most accurate way to calculate the wet-bulb temperature, "twet", is to do it in an interative fashion as there is no single formula that can determine its value. It is both a function of the temperature and the humidity; i.e. "relative humidity"/"dew point" as well as the pressure. Meteorologically speaking it is found by taking a parcel of air with a certain temperature and dew point and rising it adiabatically. Doing so, it will cool due to the reduction in pressure as it rises, allowing the particle to expand. This happens at 9.8 degrees C/km. It maintains its moisture content and, given that the amount of water vapor that can be contained in air is limited by its temperature, eventually it will reach saturation; i.e. 100% RH and cloud will form. This occurs at the condensation temperature, with the associated pressure being the "conditional level". Here the "twet" value is the same as the temperature. One then brings this parcel back down to the original pressure whilst maintaining 100% RH, i.e. conserving its "potential wet bulb temperature". This is why snow can occur at temperatures above zero, which I will address when I comment on the rain/snow topic later.

There is however, a statistical formula that has been derived that allows for a quick computation of twet close to sea-level. It is thanks to a paper by Roland Stull entitled "Wet-bulb Temperature from Relative Humidity & Air Temperature" which appeared in The November 2011 edition of the "Journal of Applied Meteorology and Climatology". Vol 50; pp 2267-2269.
[https://journals.ametsoc.org/view/journals/apme/50/11/jamc-d-11-0143.1.xml](url)

The formula is, using "R" coding but it should be self explanatory:

`twet = temp*atan(0.151977*sqrt(RH+8.313659)) + atan(temp+RH)-atan(RH-1.67631) + 0.00391838*RH^1.5*atan(0.023101*RH) - 4.686035`

where 'atan' is the inverse-tangent command, given in radians and NOT degrees, the temperature is in celsius and the humidity is in %, i.e. 0-100.  Simply plug in a value for the "temp" and the "RH". For example, temp = 4 and RH = 65, same as tdew = -2, gives twet = +1.6

As an aside, Stull is the author of a "libretext" book on meteorology called "Practical Meteorology". It, and considerably more books on other topics, are free. Personally I donate to them as they are worthy recipients. I strongly recommend anybody here to look at Stull's effort. Not only is it extremely comprehensive, he gives plenty of practical examples [hence the title] with countless worked examples. Users of Patrick's resource here, himself certainly included, might find some opportunities to use the data here for some extra ideas. The link to libretext is: [https://libretexts.org/](url) with the direct link to his book being: [https://commons.libretexts.org/book/geo-9525](url)

However, this formula only deals with sea-level and presumably Patrick and other users would prefer something without restrictions. To do this requires some coding. Back in 2006 I grabbed some code off the internet that contained FORTRAN routines to calculate various thermodynamic variables. I've just located it: [https://wahiduddin.net/calc/refs/density_algorithms.pdf](url) It is potential "gold" for anybody who wants to get deep into the data here. Included in amongst the code is a routine to calculate the wet-bulb temperature. It relies on some other subroutines. I'm not much of a programmer these days but I have easily converted the relevant routines into 'R' code and placed at the bottom of this post. Anybody who uses R should simply copy and paste and voila. All you need to input are the temperature, dewpoint and pressure. From the few calculations I have done it is pretty quick, so it might be possible for Patrick to implement, if not, those using the data here can employ it themselves.

I've given a plot below comparing the twet values derived from Stull's work with those of the adapted FORTRAN routine. There is some disagreement when the temperatures drop below zero which Stull mentions in his work. From the viewpoint of rain/snow this is irrelevant as the crucial values of twet to delineate between the 2 precipitation types is centred slightly above zero and values this far below zero won't be rain. (Could be freezing rain?)

As an aside, I continue to note that the EC hourly data I've obtained doesn't contain any dewpoints, only temperature and relative humidity. The other data sets do. Is there a reason for this? It is easy to calculate them based on the temperature and relative humidity values so I have no problems with this but I wonder if for consistency sake that Patrick can quickly derived them from the other two values. I've included an R routine to do so at the bottom of the code below.

Dave

![twet_comparison_plot](https://github.com/user-attachments/assets/43692c07-bda7-4031-b99a-f0492ccd49f0)

R code.............................. without any commentary as one can find out more in the PDF referred to above. Apologies for the formatting, I can't seems to get the code to print the way I'd like....

Main routine
============

`
twet <- function(T,Td,p) {
    aw <- w(Td,p)
    ao <- o(T,p)
    p_iter <- p; x <- 1

    while(abs(x) >= 0.01) {
        x <- .02*(tmr(aw,p_iter)-tda(ao,p_iter))
        p_iter <- p_iter*(2^(x))
      }

    tequiv <- tda(ao,p_iter)
    aos <- os(tequiv,p_iter)
    twet <- tsa(aos,p)
    round(twet,2)
  }
`

Subroutines required within "twet"

`
esat <- function(T) {

    TK <- T+273.15
    P1 <- 11.344-0.0303998*TK
    P2 <- 3.49149-1302.8844/TK
    C1 <- 23.832241-5.02808*log(TK,10)
    esat <- 10.^(C1-1.3816E-7*10.**P1+8.1328E-3*10.**P2-2949.076/TK)
    esat
  }

w <- function(T,P) {

    X <- esat(T)
    w <- 622.*X/(P-X)
    w
  }

o <- function(T,P) {

    TK <- T+273.15
    OK <- TK*((1000./P)^0.286)
    o <- OK-273.15
    o
  }

tmr <- function(W,P) {

    C1 <- .0498646455; C2 <- 2.4082965; C3 <- 7.07475; C4 <- 38.9114; C5 <- 0.0915; C6 <- 1.2035

    X <- log(W*P/(622.+W),10)
    TMRK <- 10^(C1*X+C2)-C3+C4*((10.^(C5*X)-C6)^2.)
    tmr <- TMRK-273.15
    tmr
  }

tda <- function(O,P) {

    OK <- O+273.15
    TDAK <- OK*((P*.001)^0.286)
    tda <- TDAK-273.15
    tda
  }

os <- function(T,P) {

    TK <- T+273.15
    OSK <- TK*((1000./P)^0.286)*(exp(2.6518986*w(T,P)/TK))
    os <- OSK-273.15
    os
  }

tsa <- function(OS,P) {

    A <- OS+273.15; TQ <-  253.15
    D <-  120.; X <- 1

    while(abs(X) >= 1E-7) {
        TQK <- TQ-273.15
        D <- D/2
        X <- A*exp(-2.6518986*w(TQK,P)/TQ)-TQ*((1000./P)^0.286)
        TQ <- TQ+D*sign(X)
      }
    tsa <- TQ-273.15
    tsa
  }
`

The examples for twet given in the PDF article are given below. The difference between the values obtained with this code and those given are minimal, though the last example differs a bit. Any users might wish to tweek the code with regards to the criteria for exiting of the looping. This might result in a trade-off with the time to do the calculations if one was to employ it on a large scale.

twet(30,15,1000)	# 19.99
twet(0,-20,700)	# -6.05
twet(10,0,850)	# 5.11
twet(-15,-25,500)	# -17.39
twet(25,20,900)	# 21.57

Routine for calculating the dewpoint from the temperature [deg C] & relative humidity [0-100]

`TD <- function(T,RH) {

    ES <- 6.11*10^(7.5*T/(237.3+T))
    E <- ES*RH/100

    TD <- round(237.3*log(ES*RH/611)/(7.5*log(10)-log(ES*RH/611)),1)
    TD
  }
`

**Comment by patrick-zippenfenig** [comment]
Thanks for the suggestion. The API already support `wet_bulb_temperature_2m`, but it was never added to the documentation. It is using the same formula. 

https://github.com/open-meteo/open-meteo/blob/2a3e2d67f7ce4b4fddcf0592f3190dbf1ed277c4/Sources/App/Helper/Meteorology.swift#L230-L237

I [just added it](https://github.com/open-meteo/open-meteo-website/pull/340) to the documentation and you can see a preview here: 

https://wet-bulb.open-meteo-website.pages.dev/en/docs#hourly=wet_bulb_temperature_2m&models=ecmwf_ifs04,ecmwf_ifs025,ecmwf_aifs025,cma_grapes_global,bom_access_global,gfs_seamless,jma_seamless,icon_seamless,gem_seamless,meteofrance_seamless,metno_seamless,metno_nordic,knmi_seamless,dmi_seamless,ukmo_seamless
